config/types: add support for links

dgonyeo opened the pull request:
This commit adds support for links, both symbolic and hard, to the
config as a new object at the same level as files and directories. As
links don't have their own permissions, this commit also moves the mode
field out of the node object and adds it directly to files and
directories.


Created to be an alternative solution to the issue https://github.com/coreos/ignition/pull/347 is attempting to fix.

Related: https://github.com/coreos/bugs/issues/1873

lucab commented:
Alternative schema suggestions, following nomenclature from [here](https://serde.rs/enum-representations.html):
 * External tag - move the symlink section one level up and make it an either/variant of the content section (they don't share any property anyway).
 * Internal tag - add a "kind" field to actually specify which fields apply to which case.

I *may* have a usecase for this, so :+1: from my side even if I'd prefer it organized with an external explicit tag. FWIW I don't care about hardlink, but somebody may also want to have them (in case you want to future-proof the schema).

dgonyeo commented:
rebased this down to 1 commit (json-schema was merged)

dgonyeo commented:
Updated this to have a `kind` field in the contents section, that can be either `file` or `symlink`, with an empty string defaulting to `file`.

Also added in a validation function for the `symlinkTo` field to ensure it's an absolute path.

dgonyeo commented:
review bump

crawford commented:
I don't really like lumping this under the file section. It sounds like there was an alternate proposal. What did that look like? My thinking is that since directories have their own section, so should links.

dgonyeo commented:
Yeah that seems reasonable. I'll move it to its own section.

dgonyeo commented:
links are now in their own section in the schema, and both symbolic and hard links are supported. I've update the PR to reflect this.

crawford commented:
Please break this into multiple commits.

crawford commented on doc/configuration-v2_1-experimental.md:
```diff
@@ -64,6 +64,15 @@ The Ignition configuration is a JSON document conforming to the following specif
       * **_id_** (integer): the group ID of the owner.
+  * **_links_** (list of objects): the list of links to be created
+    * **filesystem** (string): the internal identifier of the filesystem in which to write the link. This matches the last filesystem with the given identifier.
+    * **path** (string): the absolute path to the link
```
Symlinks can have relative paths.

dgonyeo replied:
This is the path where the link file is created, not what it is pointing to

crawford commented on doc/configuration-v2_1-experimental.md:
```diff
@@ -64,6 +64,15 @@ The Ignition configuration is a JSON document conforming to the following specif
+    * **_group_** (object): specifies the group of the owner.
+      * **_id_** (integer): the group ID of the owner.
+    * **target** (string): the target path of the link
+    * **_symbolic_** (boolean): a hard link is created if this is false, a symbolic one if this is true.
```
We probably want to default to symbolic.

dgonyeo replied:
Renamed to `hard` so that the zero-value means it default so symbolic.

dgonyeo commented:
Split into 3 commits.

crawford commented on config/translate.go:
```diff
@@ -89,11 +89,11 @@ func TranslateFromV1(old v1.Config) types.Config {
 					Group:      types.NodeGroup{ID: oldFile.Gid},
 				},
 				FileEmbedded1: types.FileEmbedded1{
+					Mode: int(oldFile.Mode),
```
All of these schema changes should be a separate commit (or part of the previous).

dgonyeo replied:
Squashed commits 1 and 2

dgonyeo commented:
Tweaked the PR again into the following commits:
- schema modification for the moving of `mode` out of `node`
- implementation of the moving of `mode`
- schema and docs modification for the implementation of links
- implementation of links
